x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/steps/blocks/action/math/math_expression_editor.tsx

```diff
@@ -0,0 +1,295 @@
+import { CodeEditor } from '@kbn/code-editor';
+import { i18n } from '@kbn/i18n';
+import { useController } from 'react-hook-form';
+import { LanguageDocumentationPopover } from '@kbn/language-documentation';
+import { Markdown } from '@kbn/shared-ux-markdown';
+import { monaco } from '@kbn/monaco';
+import { validateMathExpression, getMathExpressionLanguageDocSections } from '@kbn/streamlang';
+import type { ProcessorFormState } from '../../../../types';
+import { useEnrichmentFieldSuggestions } from '../../../../../../../hooks/use_field_suggestions';
+// Import to register the math language with Monaco
+import { STREAMS_MATH_LANGUAGE_ID } from './math_expression_tokenization';
+import {
+  registerMathCompletionProvider,
+  registerMathSignatureHelpProvider,
+} from './math_expression_completion';
+
+export const MathExpressionEditor: React.FC = () => {
+  const { euiTheme } = useEuiTheme();
+  const [isWordWrapped, setIsWordWrapped] = useState(true);
+  const [isHelpOpen, setIsHelpOpen] = useState(false);
+  const editorRef = useRef<monaco.editor.IStandaloneCodeEditor | null>(null);
+  const modelRef = useRef<monaco.editor.ITextModel | null>(null);
+
+  // Get field suggestions for autocompletion
+  const fieldSuggestions = useEnrichmentFieldSuggestions();
+
+  const { field } = useController<ProcessorFormState, 'expression'>({
+    name: 'expression',
+    rules: {
+      required: i18n.translate('xpack.streams.math.editorExpressionRequired', {
+        defaultMessage: 'An expression is required.',
+      }),
+      validate: (value: string) => {
+        if (!value.trim()) return true;
+        const result = validateMathExpression(value);
+        if (!result.valid) {
+          return result.errors.join('; ');
+        }
+        return true;
+      },
+    },
+  });
+
+  // Real-time validation for inline display (separate from form validation)
+  const validationResult = useMemo(() => {
+    if (!field.value || !field.value.trim()) return null;
+    return validateMathExpression(field.value);
+  }, [field.value]);
+
+  const hasInlineError = validationResult && !validationResult.valid;
+
+  // Set error markers on the Monaco model for squiggly underlines
+  useEffect(() => {
+    const model = modelRef.current;
+    if (!model) return;
+
+    if (hasInlineError && validationResult) {
+      // Create markers for each error
+      // Since we don't have precise position info, mark the entire content
+      const markers: monaco.editor.IMarkerData[] = validationResult.errors.map((error) => ({
+        severity: monaco.MarkerSeverity.Error,
+        message: error,
+        startLineNumber: 1,
+        startColumn: 1,
+        endLineNumber: model.getLineCount(),
+        endColumn: model.getLineMaxColumn(model.getLineCount()),
+      }));
+      monaco.editor.setModelMarkers(model, 'math-validation', markers);
+    } else {
+      // Clear markers when valid
+      monaco.editor.setModelMarkers(model, 'math-validation', []);
```

Comment:
You might want to also add a cleanup function for when the component unmounts, I'm not sure if it really matters but just incase Monaco tries to do something with them in the lifecycle:

```ts
return () => {
    if (modelRef.current) {
      monaco.editor.setModelMarkers(modelRef.current, 'math-validation', []);
    }
  };
```

Reply:
Yes, it should clear in useEffect's cleanup. Thanks for spotting this.

Will do as a follow up to not risk running the CI cycle on FF day. 

Currently it clears the markers on initialization. Since 'math-validation' model is only used for for this component, it works but its's not future proof.